nexus/src/db/datastore/resource_usage.rs

```diff
@@ -0,0 +1,221 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! [`DataStore`] methods on [`ResourceUsage`]s.
```

Comment:
I think more comments on the design here would be worthwhile.  It's taking me a while to understand this.  I guess we have an Oximeter producer that's emitting data points to Oximeter to go into Clickhouse.  Presumably this is the mechanism by which the resource usage metrics that we're tracking in CockroachDB get to Clickhouse?  But this Nexus instance only knows _some_ of the usage -- so is this only responsible for emitting deltas?

Am I reading the code right that when Nexus does an operation that changes the allocation numbers, it _separately_ does two things: (1) updates the table and (2) emits a data point to Oximeter?

Reply:
Added a comment in 9a758c2b0612ba692664ca66e4141beed32af4d0 .

- Whenever `ResourceUsage` objects are created or modified, we append samples to this Producer
- CRDB always holds the "known" totals. Clickhouse is informed of these samples as they are collected by the Oximeter collector. This producer just acts as a mechanism for batching and sending them as a group periodically.
- It's true that this Nexus instance only may receive *some* of the operations. However, storing these data points is more of a batching mechanism than anything else - if three different nexus instances updates resource usages separately, they should be conjoined once collected into Clickhouse.
- All samples emitted to Clickhouse are "absolute values" -- nothing relies on the diff of what happened to exist in Clickhouse beforehand.

Reply:
Okay.  Your GitHub comment above has detail that would be useful in the code comment I think.

I'm still not sure how this could be correct if each Nexus only sees some of the updates but reports absolute sums to Clickhouse.  Would it be worth working through an example where two different Nexus instances each create a Disk (different sizes, for clarity)?  How does Clickhouse wind up with the right total for the Project, Organization, and Silo?